### ISSUE TITLE
On the topic of JSON serialization roundtrips

### ISSUE THREAD

**Issue by Stephan202** [immutables/immutables#464]
I'd like to start the discussion with a test class; tried to keep it concise:

```java
package com.example;

import static org.testng.Assert.assertEquals;

import com.fasterxml.jackson.annotation.JsonAutoDetect;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.PropertyAccessor;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import java.io.IOException;
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;
import org.immutables.value.Value;
import org.testng.annotations.Test;

@Test
public final class ImmutablesTest {
    @JsonSerialize
    @JsonDeserialize
    @Value.Immutable
    public abstract static class Dummy {
        @Value.Parameter
        public abstract boolean isSomeProperty();
    }

    @JsonSerialize
    @JsonDeserialize
    @Value.Immutable
    public abstract static class CustomDummy {
        @Value.Parameter
        @JsonProperty("custom_name")
        public abstract boolean isSomeProperty();
    }

    @JsonSerialize
    @JsonDeserialize
    @Retention(RetentionPolicy.CLASS)
    @Target({ElementType.PACKAGE, ElementType.TYPE})
    public @interface Model {}

    @Model
    @Value.Immutable
    public abstract static class DummyWithMetaAnnotation {
        @Value.Parameter
        public abstract boolean isSomeProperty();
    }

    @Model
    @Value.Immutable
    public abstract static class CustomDummyWithMetaAnnotation {
        @Value.Parameter
        @JsonProperty("custom_name")
        public abstract boolean isSomeProperty();
    }

    public void testDummyWithDefaultObjectMapper() throws IOException {
        verifyRoundTrip(getMapper(false), ImmutableDummy.of(true));
    }

    public void testDummyWithCustomObjectMapper() throws IOException {
        verifyRoundTrip(getMapper(true), ImmutableDummy.of(true));
    }

    public void testCustomDummyWithDefaultObjectMapper() throws IOException {
        verifyRoundTrip(getMapper(false), ImmutableCustomDummy.of(true));
    }

    // Fails: Unrecognized field "isSomeProperty" (class com.picnic.fulfillment.model.ImmutableCustomDummy$Json), not marked as ignorable (one known property: "custom_name"])
    public void testCustomDummyWithCustomObjectMapper() throws IOException {
        verifyRoundTrip(getMapper(true), ImmutableCustomDummy.of(true));
    }

    // Fails: Unrecognized field "someProperty" (class com.picnic.fulfillment.model.ImmutableDummyWithMetaAnnotation$Json), not marked as ignorable (one known property: "isSomeProperty"])
    public void testDummyWithMetaAnnotationWithDefaultObjectMapper() throws IOException {
        verifyRoundTrip(getMapper(false), ImmutableDummyWithMetaAnnotation.of(true));
    }

    public void testDummyWithMetaAnnotationWithCustomObjectMapper() throws IOException {
        verifyRoundTrip(getMapper(true), ImmutableDummyWithMetaAnnotation.of(true));
    }

    // Fails: Unrecognized field "custom_name" (class com.picnic.fulfillment.model.ImmutableCustomDummyWithMetaAnnotation$Json), not marked as ignorable (one known property: "isSomeProperty"])
    public void testCustomDummyWithMetaAnnotationWithDefaultObjectMapper() throws IOException {
        verifyRoundTrip(getMapper(false), ImmutableCustomDummyWithMetaAnnotation.of(true));
    }

    // Fails: Unrecognized field "custom_name" (class com.picnic.fulfillment.model.ImmutableCustomDummyWithMetaAnnotation$Json), not marked as ignorable (one known property: "isSomeProperty"])
    public void testCustomDummyWithMetaAnnotationWithCustomObjectMapper() throws IOException {
        verifyRoundTrip(getMapper(true), ImmutableCustomDummyWithMetaAnnotation.of(true));
    }

    private ObjectMapper getMapper(final boolean useFields) {
        final ObjectMapper mapper = new ObjectMapper();
        return useFields
                ? mapper.setVisibility(PropertyAccessor.ALL, JsonAutoDetect.Visibility.NONE)
                        .setVisibility(PropertyAccessor.FIELD, JsonAutoDetect.Visibility.ANY)
                : mapper;
    }

    private void verifyRoundTrip(final ObjectMapper mapper, final Object value) throws IOException {
        final String json = mapper.writeValueAsString(value);
        final Object newValue = mapper.readValue(json, value.getClass());
        assertEquals(newValue, value);
    }
}
```

This test defines four immutable models and eight tests, two for each model:

- Two of the models are defined with direct `@JsonSerialize`/`@JsonDeserialize` annotations, while the other two inherit said annotations from a meta annotation.
- For each of the above two variants, there is a model with and without a custom `@JsonProperty` annotation.
- For each of the four models we perform two serialization round trip tests: once with the standard Jackson `ObjectMapper`, and once with an `ObjectMapper` which looks at field names rather than method names.

Four out of the eight tests fail, all because Jackson deserialization fails. I added a comment with the error message above each of these tests.

My questions:
- Which of these examples do you think should work?
- Can we update Immutables such that all work?

Reason I bring this up, is that we want to use Immutables in a Spring MVC application with a lot of model classes, in which a custom object mapper is used. (It's similar to the one above which looks at field names). Migrating to Immutables will be a gradual process, and so the requirements are that:

- Any models we generate using Immutables are compatible with the existing `ObjectMapper`, yielding JSON with field names that precisely match those of the model classes being replaced.
- Customizations to the models generated with Immutables (such as `@JsonSerialize`/`@JsonDeserialize`) are defined on a meta annotation, so as to avoid repetition.

I think the first issue can be fixed by making sure that `@JsonProperty` annotations are not only placed on generated accessor methods, but also on the mutator methods of the generated `Json` class. I'm not yet sure why the meta-annotation approach doesn't "just work".

Would love to hear your thoughts :)

**Comment by elucash** [comment]
Thank you for raising this! I've found a bug in meta-annotation code where `@JsonProperty`s were not generated in a `Json` class. When tested with the fix, only 2 test are still failing

```
testCustomDummyWithCustomObjectMapper
testCustomDummyWithMetaAnnotationWithCustomObjectMapper
```

Of course, the reason for the fail is additional JSON fields derived from internal fields.

For better or for worse, Immutables were designed with the notion of data hiding built-in. On one hand, we're trying to maintain simplest possible semantic, as objects are immutable after all, on the other hand, we do have places where there are reasons why we don't have 1:1 correspondence between attributes and fields: derived, lazy, optional (the one from `java.util`) have some additional field trackers fields or have different field types. It's getting even more complicated with the creation of [encodings](https://github.com/immutables/immutables/issues/363) which allow customization of how attributes are implemented, including having arbitrary backing fields.

Supporting fields following the exact the same representation as attributes is not an option. And I believe in Jackson, objects mapped using fields could coexist with the ones that are mapped by accessors, but not in the way forced by `objectMapper.setVisibility`.

I don't know how exactly fix this, but I suggest trying to change the way object mapper is initialized. On the side of Immutables, we just cannot disable reflection or guarantee 1:1 correspondence for fields and accessors. In the next release, there will new style option to disable any implicit Jackson integration (`Style(jacksonIntegration = false)`). It might help in some cases where default integration is not doing the right thing.

**Comment by Stephan202** [comment]
Hi @elucash. Thanks for fixing the meta-annotation support!

About supporting field serialization: this is actually _not_ what I'm suggesting :). Already, Immutables supports adding `@JsonProperty` annotations on methods, and as the error messages of the two remaining failing tests show, Jackson will respect those annotations even when the custom object mapper is used.

Admittedly, if Immutables generated additional fields under the hood, then those will by default be serialized as well. But that can be avoided by marking those fields `transient` or annotating them with `@JsonIgnore`.

Anyway, let's focus on the simple case, in which no additional fields are generated. Then, we have the serialization part covered by adding `@JsonProperty` to the method declarations. My question is: can we add _mirroring_ annotations on the various `ImmutableType.Json#setXXX` methods? That should get the other two tests to pass as well, and basically allow one to use Immutables with any kind of object mapper visibility configuration.

NB 1: The reason I'm not just opening a PR for this is precisely because of those extra features that may complicate things: I'd like to discuss the design space with you first. Also, while for getters one may want to copy over all annotations defined on the interface/abstract class, for the setters one will probably want to cherry-pick just the `@JsonProperty` annotations... I think. (B.t.w., optionals shouldn't be a problem, assuming one registers [jackson-datatype-jdk8](https://github.com/FasterXML/jackson-datatype-jdk8).)

NB 2: I agree that simply changing the object mapper would be better, but as stated above, I'm trying to migrate a large project to Immutables, piece by piece. Changing the object mapper would have a major impact on hundreds of other model classes. So the only way for me to move forward safely, is to find a way to make Immutables play nicely with the existing config. Once we've migrated... who knows ;).

Thanks for thinking along!

**Comment by elucash** [comment]
Thank you for understanding! I appreciate your concerns about gradual migration, just was curious about if there's a way to mark different "visibility" strategies for a different set of object types while being handled by the same `ObjectMapper`.

> My question is: can we add mirroring annotations on the various ImmutableType.Json#setXXX methods?

I'm not quite got the _mirroring_ part, as I've thought we had this covered, just as an example taken from one of the tests:

```java
  @Deprecated
  @JsonDeserialize
  @JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.NONE)
  static final class Json implements DefaultCollection {
    List<String> nullable = null;
    List<String> defaults = ImmutableList.of();
    boolean defaultsIsSet;
    List<Integer> defnullable = null;
    boolean defnullableIsSet;
    @JsonProperty("nullable")
    public void setNullable(@Nullable List<String> nullable) {
      this.nullable = nullable;
    }
    @JsonProperty("defaults")
    public void setDefaults(List<String> defaults) {
      this.defaults = defaults;
      this.defaultsIsSet = true;
    }
    @JsonProperty("defnullable")
    public void setDefnullable(@Nullable List<Integer> defnullable) {
      this.defnullable = defnullable;
      this.defnullableIsSet = true;
    }
    @Override
    public List<String> nullable() { throw new UnsupportedOperationException(); }
    @Override
    public List<String> defaults() { throw new UnsupportedOperationException(); }
    @Override
    public List<Integer> defnullable() { throw new UnsupportedOperationException(); }
  }
```

We are copying over `@JsonProperty` annotations, everything else is, supposedly, being handled by  inheritance and merging of an attribute metadata performed by Jackson.  I guess, we can fix remaining issues once we can nail them down.

Let's somehow try to distill what we can do on the side of Immutables to accommodate the use case, at the same time not defeating the purpose (the concerns explained in the previous comments). Currently, there's a probability of using `@JsonIgnore` on fields (`transient` will mess with the binary serialization if someone uses it). What else we can do about it?

**Comment by elucash** [comment]
@Stephan202 Thank you for the PR! I've added the test based on the one you provided here and also updated `Renderers.generator`. This will see the light in the next version, I hope to release it promptly, just want to add a couple of additional fixes from the other issues.

**Comment by elucash** [comment]
Released in 2.3.7
Important note: I had to add `forceJacksonIgnoreFields` style flag to enable this. Apparently, it's causing unexpected conflicts with some other Jackson functionality (#469)
